### ISSUE TITLE
[RFC] Consider deprecating & replacing `DataTypes.DATE`

### ISSUE THREAD

**Issue by ephys** [sequelize/sequelize#14295]
`DataTypes.DATE` is a bit of a mess. It has very different meanings depending on the dialect. Some implementations include a timezone, others do not.
It doesn't provide a way to choose whether the type should include a timezone.
Finally, `DATE` means `DATEONLY` in SQL, not `DATETIME`.

I propose we deprecate that DataType and replace it with:

- `DataTypes.DATETIME`: Always means "Date without timezone".
- `DataTypes.DATETIME.ZONED`: Always means "Date with timezone", throws if the dialect doesn't support this.

**Comment by mytharcher** [comment]
My proposal for the naming:

* `DateTypes.DATETIME`: Always means "Date **with** timezone". Because it will work fine in most cases by default.
* `DataTypes.DATETIME.PLAIN`: Always means "Date **without** timezone". It could only accepts `Temporal.PlainDateTime`.

Many developers arguing in other issues for demanding of date/time without timezone, because they didn't find out the best practice of usage of time.

In my opinion, when we want to use a date/time field in database, it should always be with timezone. And physically, time is always with timezone. The cases when people are not mentioning timezone, mostly they only consider the local time. But use time with timezone in these cases are also acceptable and no bad effects.

The base reason why should we use date/time type of fields in database, is Just because we need them to be calculate with other values (comparing, manipulating, etc.). If people only need time for literally expressing, the timezone will not be strictly necessary. In this case, just use plain time or even string will be ok. But if we use time for calculating, we need the exact value (with timezone).

Another thing is, always use UTC formatted in any transferring is the best practice. Also for storing to database. So in PG it works great by default. It will be the clients' stuff to transform stored time to local timezone in view. And by default `Date` in browser parsing an UTC time string works right.

Think about it, if we use time with timezone in a project, whether if it is a local one or global, it always work fine. But if use time without timezone, it only works for local projects.

**Comment by crazyscience** [comment]
> But use time with timezone in these cases are also acceptable and no bad effects.

Super not true. When you have a table with 30 billion rows and 4 datetime columns in each row, the addition of useless data (in this case, timezone info) definitely has an impact. Maybe not noticeable when someone's building yet another todo app. But ORMs should be built for performance application and production loads, not hobbyists building a blog.

Additionally, timezone boundaries move constantly. Storing the timezone data in the database as if it's going to be accurate for any length of time is seriously misguided. Production applications should **default** to avoid timezones and use them when they are needed specifically.

https://www.youtube.com/watch?v=-5wpm-gesOY

**Comment by crazyscience** [comment]
My proposal for the naming:
- `DateTypes.DATETIME`: Always means "Date without timezone". Because it will work fine in most cases by default.
- `DataTypes.DATETIME.IM_A_MASOCHIST`: Always means "Date with timezone". :smile:

**Comment by mytharcher** [comment]
@crazyscience 

You may have a check on those DBs docs, whether the DATE type with timezone or without, it will not affect size of storing.

* PG: <https://www.postgresql.org/docs/current/datatype-datetime.html>
  * Timestamp with timezone: 8 Bytes
  * Timestamp Without timezone: 8 Bytes
* MySQL: <https://dev.mysql.com/doc/refman/8.0/en/storage-requirements.html>
  * DATETIME: 8 Bytes

You can always use UTC time to store into DB, the timezone will be calculated or not on Sequelize or client side. Unless you just need local time, and no need to convert, you just store local time. The size of each has no difference.

For another aspect, performance optimization should be the App developers' responsibility. As for framework/library developer, should just provide the options. You can also use the `DataTypes.DATETIME.PLAIN` for no conversion needed fields. It all by your own.

**Comment by crazyscience** [comment]
@mytharcher while I concede that the storage does indeed appear to be unaffected, I still would maintain that a framework has a lot of sway in how programming patterns are adopted and therefor still think that the default behavior should reflect best practices.

Per the postgres link you sent above, directly below the table of storage sizes, they have a remark that says that according to the SQL standards, datetime columns must not include timezone data unless explicitly stated. I still agree with the SQL standards on this one. Timezones in the database are a bit of a foot :gun:.

**Comment by ephys** [comment]
The reason I originally went for `DataTypes.DATETIME` (unzoned) + `DataTypes.DATETIME.ZONED` is that we currently have `DataTypes.TIME` (unzoned) & have a PR for `DataTypes.TIME.ZONED`.

I think mixing up which one is zoned by default and which one isn't would lead to confusion.

However, it makes sense for `DataTypes.TIME` to be unzoned by default, because, as the pg docs indicate, the usefulness of `DataTypes.TIME.ZONED` is questionable. That is not true for `DataTypes.DATETIME.ZONED`.

Whether we make zoned or unzoned the default is inevitably going to lead to people using it without thinking about whether they want it to be zoned or not. This is mitigated by the fact that we will only accept `ZonedDateTime` & `PlainDateTime` respectively, so they will eventually have to think about which type they really need, but maybe we should go the route of no default here.

- `DataTypes.DATETIME`: error, please use `DataTypes.DATETIME.ZONED`, or `DataTypes.DATETIME.PLAIN`
- `DataTypes.DATETIME.PLAIN`
- `DataTypes.DATETIME.ZONED`

or just

- `DataTypes.ZONEDDATETIME`
- `DataTypes.PLAINDATETIME` (this uppercase convention makes this somewhat difficult to read)

**Comment by WikiRik** [comment]
Yeah, I don't think `DataTypes.ZONEDDATETIME` is a good solution. I think it should be fine to have no default and force people to specify `DataTypes.DATETIME.ZONED`.

**Comment by ephys** [comment]
Something I didn't put much emphasis on is that no database currently support a type that is suitable to `Temporal.ZonedDateTime`, because the timezone & calendar extensions [have not been standardized yet](https://github.com/tc39/proposal-temporal/issues/1450). The best we have is a datetime + offset db type. We should reserve `.ZONED` for proper timezone support (let's hope that happens one day), and use `.OFFSET` instead.

I think the support table would look like this:

| Sequelize DataType          | JS Temporal Type | pg           | mysql/mariadb | mssql          |
|-----------------------------|------------------|--------------|---------------|----------------|
| `DataTypes.DATETIME.ZONED`  | ZonedDateTime    | ❌            | ❌             | ❌              |
| `DataTypes.DATETIME.OFFSET` | Instant          | timestamp tz | ❌             | datetimeoffset |
| `DataTypes.DATETIME.PLAIN`  | PlainDateTime    | timestamp    | datetime      | datetime2      |
| `DataTypes.TIME`            | PlainTime        | time         | time          | time           |
| `DataTypes.TIME.OFFSET`     | ❌                | time tz      | ❌             | ❌              |
| `DataTypes.DATEONLY`        | PlainDate        | date         | date          | date           |

That has implications for #14213. We need to name that feature `TIME.OFFSET` instead of `TIME.ZONED`

For now, I think trying to use `ZonedDateTime` with one of these types should emit an error explaining that no DB type matching `ZonedDateTime` is available yet, and that the user should
split it into two attributes: a plain datetime & a timezone

`Temporal.{PlainMonthDay, PlainYearMonth, TimeZone, Calendar}` don't have an associated JS nor DB DataType

**Comment by ephys** [comment]
Updated the proposal with the following changes:
- Removed `DataTypes.TIME.OFFSET`, see #14213 
- Added mapping for DB2 and Oracle

**Comment by ephys** [comment]
As mentioned in https://github.com/sequelize/sequelize/issues/4900, we can also map the `INTERVAL` data type to `Temporal.Duration`
